### ISSUE TITLE
Occasional wildly incorrect results from numpy.linalg.lstsq

### ISSUE THREAD

**Issue by adian0** [numpy/numpy#5702]
Hello,

I'm new to numpy and not exactly an expert in statistics.  However, I think I've run across a serious bug in numpy's linear regression. I'm running under Linux, Python3.4 on a 64bit processor and OS.  

Please see the script below.  The graphical output speaks for itself, so I encourage you to run it.  See also the comments inline.

I suspect the bug is related to an integer/floating type cast/overflow or something along along those lines.  With my data in similar ranges, it seems to occur when the x coordinates are nearer to 2**32-1, though I don't have a lot of examples.

```
#!/usr/bin/env python3

import numpy
import matplotlib.pyplot as plt

def runRegression(x,y):
    # The x1 data set fails no matter whether dtype is specified here
    x = numpy.array(x, dtype='int64') 
    y = numpy.array(y, dtype='int64')
    A = numpy.vstack([x, numpy.ones(len(x))]).T
    m, c = numpy.linalg.lstsq(A, y)[0]
    print(m,c)
    
    plt.clf()
    plt.scatter(x, y)
    plt.plot(x, m*x + c, 'r', label='Fitted line')
    plt.show()


# data set 1 (buggy result)
x1=[4155915305, 4155915503, 4155915996, 4155916186, 4155916383, 4155916572, 4155916765, 4155916956, 4155917146, 4155917336, 4155917527, 4155917717, 4155917907, 4155918097, 4155918287, 4155918477, 4155918669, 4155918859, 4155919065, 4155919255, 4155919446, 4155919637, 4155919839, 4155920029, 4155920221, 4155920411, 4155920601, 4155920791, 4155920981, 4155921171, 4155921361, 4155921711, 4155921712, 4155921713]
y1=[1426868220654354484, 1426868220851239808, 1426868221345173820, 1426868221535528709, 1426868221731000861, 1426868221919978735, 1426868222113280237, 1426868222304021709, 1426868222494154117, 1426868222684135813, 1426868222874947575, 1426868223064930705, 1426868223254993810, 1426868223444995738, 1426868223635010452, 1426868223825077194, 1426868224017034192, 1426868224207064981, 1426868224413369035, 1426868224603382903, 1426868224794319018, 1426868224985246295, 1426868225187237516, 1426868225378046156, 1426868225569093390, 1426868225759105034, 1426868225949105820, 1426868226139548300, 1426868226329221050, 1426868226519028305, 1426868226709140271, 1426868227059669014, 1426868227060485049, 1426868227061232442]

# data set 2 (correct result)
x2=[2547893056, 2547893254, 2547893754, 2547893949, 2547894142, 2547894332, 2547894522, 2547894714, 2547894904, 2547895100, 2547895290, 2547895480, 2547895670, 2547895860, 2547896054, 2547896246, 2547896436, 2547896630, 2547896820, 2547897012, 2547897202, 2547897395, 2547897591, 2547897789, 2547897979, 2547898171, 2547898363, 2547898559, 2547898749, 2547898939, 2547899129, 2547899463, 2547899463, 2547899464]
y2=[1426868220654424592, 1426868220852333793, 1426868221351208331, 1426868221546516199, 1426868221739105396, 1426868221929473989, 1426868222119220265, 1426868222311370470, 1426868222502055421, 1426868222697320176, 1426868222887046740, 1426868223077035515, 1426868223267155631, 1426868223457194795, 1426868223651223962, 1426868223843296189, 1426868224033102573, 1426868224227204330, 1426868224417226660, 1426868224609132625, 1426868224799268267, 1426868224992820750, 1426868225188368151, 1426868225386159254, 1426868225576071133, 1426868225768232901, 1426868225960498827, 1426868226156004134, 1426868226346143762, 1426868226536079731, 1426868226726248163, 1426868227060517735, 1426868227061182598, 1426868227061217040]

# Scaled versions of two data sets.  This "fixes" the issue in data set 1
# Note that simply casting x/y values to floats didn't fix the problem.
x1f = [x/1000.0 for x in x1]
y1f = [y/1000.0 for y in y1]
x2f = [x/1000.0 for x in x2]
y2f = [y/1000.0 for y in y2]


runRegression(x1,y1)
runRegression(x2,y2)
runRegression(x1f,y1f)
runRegression(x2f,y2f)

#Upon running, stdout reads:
# 343334003.527 0.0826132616779
# 999948.658089 1.42432045841e+18
# 987046.894603 1.42276613791e+15
# 991294.522819 1.42434250873e+15
#
# All slopes should be near 1000000, but first run is buggy.
# The first plot makes the bug very clear, so try running it yourself.
```

**Comment by charris** [comment]
The condition number of the failing example is about 1e16, it is expected to fail given the standard approach of lstsq to almost singular design matrices. Scaling does affect the condition number, or you can play with the `rcond` parameter, although in this case I don't expect it to help. The polynomial fitting routines do scaling, so 
```
In [11]: from numpy.polynomial import Polynomial as P

In [12]: P.fit(x1, y1, 1).convert()
Out[12]: Polynomial([  1.42271246e+18,   9.99963625e+05], [-1.,  1.], [-1.,  1.])
```

**Comment by pv** [comment]
Your fitting problem is numerically ill conditioned, calculate the
singular values of A and note their ratio is large, of the order of 10^15.
.
In other words, the relative change in your x and y coordinates is too
small for floating point. The variation in y starts after 9th digit, and
variation in x after 6th digit. 9+6=15. Double precision floating point
only has ~16 digits of precision.

**Comment by charris** [comment]
Note that the condition number of the Polynomial fit, which does offset and scaling of the polynomial domain as well as scaling of the columns of the Vandermonde matrix, is ~1.13, a big improvement.

**Comment by adian0** [comment]
Uh... I've already lost you. =P  Are you saying this isn't a bug and that I'm using lstsq incorrectly?  OR that Polynomial.fit() is a reasonable work-around to the bug?

Polynomial.fit() definitely gives the correct result with this data set, so I can use that, provided you don't think the issue will rear it's head again with slightly different data.  

Thanks for the quick reply!

**Comment by charris** [comment]
It's not a bug, it is a property of the problem as stated. Scaling does change the problem, and as a rule of thumb is something that you should be aware of and use when appropriate. Google `condition number` for more explanation, but in general, a big condition number means trouble. If you use Polynomial.fit, be aware that some precision is lost with the convert method above, but it has the virtue of returning the coefficients as you probably expect them to be. You can plot either of the unconverted polynomial `p` that results with
```
plot(*p.linspace)
```
Note also that the coefficients go from low to high degree.

**Comment by charris** [comment]
Closing now.

**Comment by argriffing** [comment]
The OP's problem looks a lot like the example in http://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.lstsq.html.  Should the documentation in that example be changed to say something like 'A more stable solution is to use `Polynomial.fit` instead.  Google `condition number` for more explanation'?

**Comment by charris** [comment]
Another thing to keep in mind is that Python integers have as many digits as needed, but floats are only good for about 16 digits. If your numbers are constant for 16 digits, you will not see any variations after floating conversion. The way to work around that is to subtract the median of the integer data to get rid of the offset.

**Comment by charris** [comment]
@argriffing The examples look good for condition number, but perhaps a mention of what can go wrong would help. Perhaps an example of how to compute the condition number from the singular values in the full output would also help. A full tutorial on least squares fitting is probably out of scope ;)

**Comment by adian0** [comment]
Ok, that makes sense why the issue comes up.  I understand integer/floating point conversion all too well.  But from a user's perspective, it is a big pitfall.  Python has one set of rules about integers/floating point, and numpy has another.  Without warnings in the documentation or at run-time when the ranges get absurd (for numpy), you're bound to have this issue come up again and again.  And in my case, most results looked fine, but just once in a while they'd get wild.  Not a good mode of failure.  A slight loss in precision with polyfit is a small price to pay for reasonable output, so long as it is documented.

I agree with @argriffing: I don't expect you to educate folks like me me on the rules and limitations of regressions, but I think some basic warnings and discussion of limitations are warranted (perhaps with links to sites with more background information).

**Comment by argriffing** [comment]
This is not an answer to the original question, but the `rank` output of `lstsq` could be a helpful diagnostic. In the script, only the first return value is used (this is unpacked into `m` and `c`).  The third return value gives the `rank` that is internally used by the least squares solver.  In the four cases evaluated by the script, the returned `rank` is 1 for the wildly incorrect case and 2 for the three other more successful cases.

**Comment by charris** [comment]
@adian0 Note that polyfit is not the same as Polynomial.fit. In particular, it doesn't offset the polynomials, which will cause lack of precision with the sort of data you have.

**Comment by adian0** [comment]
OK, good to know, thank you.  I had fallen back on polyfit because the documentation on datatypes seems to be scattered and I didn't immediately see how to get my coefficients off of the Polynomial object. But I finally figured it out, so I'm using Ploynomial.fit() now.

One suggestion I have, from a naive end user perspective, is to provide more centralized/indexed documentation on numpy's data types.  It seems like there's a ton of custom data types in the library, but it's not always easy to figure out what methods and attributes are available on them.  For instance, when some Polynomial methods return a "series", what exactly is that?  Where are the docs for the data type?  In another example, the Polynomial class page mentions:  "The Polynomial class provides [...] the attributes and methods listed in the ABCPolyBase documentation."  Ok, so where is ABCPolyBase?  I still haven't found it.  There's no mention of the ".coef" attribute on the Polynomial page, so I was stuck until I found a separate page of examples where you guys show it being used.  Narrative examples pages are great for starting points and common tasks, but as soon as you move beyond that, being able to navigate between datatype pages becomes pretty important.  Also, I tried just looking at pydoc3 to see what attributes it showed, but the tool blew up when I dug into the polynomial modules.

Anyway, just wanted to provide feedback.  Thanks again for you help.
